File: py/portfolio.py
from __future__ import annotations
import functools
import operator

from bank import Bank
from money import Money


class Portfolio:
    def __init__(self) -> None:
        self.moneys = []

    def add(self, *moneys):
        self.moneys.extend(moneys)

    def evaluate(self, bank: type[Bank], currency: str):
        total = 0.0
        failures = []
        for m in self.moneys:
            try:
                total += bank.convert(m, currency).amount
            except Exception as ex:
                failures.append(ex)

        if not len(failures):
            return Money(total, currency)

        failure_message = ",".join(f.args[0] for f in failures)
        raise Exception("Missing exchange rate(s):[{0}]".format(failure_message))

File: js/portfolio.js
const Money = require('./money');

class Portfolio {
  constructor() {
    this.moneys = [];
  }

  /**
   * Add money to portfolio
   * @param  {...Money} moneys
   */
  add(...moneys) {
    this.moneys = this.moneys.concat(moneys);
  }

  /**
   * Evaluae the total availabe in portfolio in required Currency
   * @param {Bank} bank
   * @param {string} currency
   * @returns {Money}
   */
  evaluate(bank, currency) {
    let failures = [];
    const total = this.moneys.reduce((sum, money) => {
      try {
        const convertedMoney = bank.convert(money, currency);
        return sum + convertedMoney.amount;
      } catch (error) {
        failures.push(error.message);
        return sum;
      }
    }, 0);
    if (!failures.length) {
      return new Money(total, currency);
    }
    throw new Error(`Missing exchange rate(s):[${failures.join()}]`);
  }
}

module.exports = Portfolio;
